### PR TITLE
Add loading and running recipes to the notebook API

### LINKED ISSUE #953: Ambiguous cache retrieval in regrid preprocessor
**Describe the bug**

In #907 I came across the following test failure:

```
self = <tests.unit.preprocessor._regrid.test_regrid.Test testMethod=test_regrid__cell_specification>

    def test_regrid__cell_specification(self):
        specs = ['1x1', '2x2', '3x3', '4x4', '5x5']
        scheme = 'linear'
        for spec in specs:
            result = regrid(self.src_cube, spec, scheme)
            self.assertEqual(result, self.regridded_cube)
>           self._check(spec, scheme, spec=True)

tests/unit/preprocessor/_regrid/test_regrid.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/unit/preprocessor/_regrid/test_regrid.py:23: in _check
    self.assertEqual(_CACHE[spec], self.tgt_grid)
E   AssertionError: <iris 'Cube' of unknown / (unknown) (latitude: 180; longitude: 360)> != <Mock spec='Cube' id='140424322565072'>
```

I could not reproduce the error on my machine, and it occasionally passed on CircleCI. I tracked it down to being related to the `_CACHE`:

https://github.com/ESMValGroup/ESMValCore/blob/de112e449c2aa8fb1d132816c8c5cd7a3186b821/esmvalcore/preprocessor/_regrid.py#L42

The cache stores the results of the regridding as a simpley key, `1x1`, `2x2`, etc. The problem is that the key assignment is ambiguous. This means after multiple calls, there _will_ be a name collision, so that the wrong cached value is retrieved. This may lead to data corruption, and in my case, random test failures. 

For #907, I believe that the outcome of the test depends on whether the regridder goes first (pass) or last (second). I happened to catch this, because I implement additional tests which also call the regridder at some point.

I have worked around the issue the issue by clearing `_CACHE` in #907 , but a less naive implementation of the cache is necessary.

### PR DESCRIPTION
**Pull request by stefsmeets**
Hi everyone, this PR implements an api to load and run recipes in an interactive jupyter notebook.

This PR is mostly complete in terms of what it is supposed to do. The thing that is missing is to have good test. I'm thinking we should have some very basic test recipes that we can run to test the API.

Also implements bugfix that closes #953

**To do**
- [x] Look at issue with multi-processing (`max_parallel_tasks != 1`) -> Seems to work as intended.
- [x] Implement unit tests with basic recipe for testing (`.yml` or maybe pass dict directly?)
- [x] https://github.com/ESMValGroup/ESMValTool_sample_data/pull/15 merged
- [x] https://github.com/ESMValGroup/ESMValTool_sample_data/pull/16 merged and release/depend on v0.0.3

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Follows up on #901 and #868 
For more discussion, see issue #498 
Documentation: https://esmvaltool--907.org.readthedocs.build/projects/ESMValCore/en/907/api/esmvalcore.api.html 
Demo notebook: https://gist.github.com/stefsmeets/1163eeea765906550f772c15b69e866e

### CODE REVIEW

**Comment by stefsmeets** [comment]
Hi @nielsdrost and @Peter9192 , this PR is ready for review. Could you have a look?

There is still a small issue with the tests on CircleCI that I will have a look at, but I think other than that the PR can be reviewed. The tests are running fine locally.

**Comment by stefsmeets** [comment]
Tests are fixed, just one unrelated to this PR 👍

```
self = <tests.unit.preprocessor._regrid.test_regrid.Test testMethod=test_regrid__cell_specification>

    def test_regrid__cell_specification(self):
        specs = ['1x1', '2x2', '3x3', '4x4', '5x5']
        scheme = 'linear'
        for spec in specs:
            result = regrid(self.src_cube, spec, scheme)
            self.assertEqual(result, self.regridded_cube)
>           self._check(spec, scheme, spec=True)

tests/unit/preprocessor/_regrid/test_regrid.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/unit/preprocessor/_regrid/test_regrid.py:23: in _check
    self.assertEqual(_CACHE[spec], self.tgt_grid)
E   AssertionError: <iris 'Cube' of unknown / (unknown) (latitude: 180; longitude: 360)> != <Mock spec='Cube' id='140715029656144'>
```

File: esmvalcore/experimental/utils.py
Comment (Peter9192) [inline esmvalcore/experimental/utils.py:99]: Perhaps it would be nice to print here something like "Found recipe_blablalba.yml" in "interesting/path". In that way, we could avoid any confusion as to which recipe is being used if there happen to be two of them when only 1 is expected or vice versa.

**Review by Peter9192 — COMMENTED** [review]
Hey Stef!
Very nice work so far!! I had a first look at the code - didn't test anything yet. But I have a couple of suggestions and questions. My main concern is that I'm a bit worried that the distinction between RecipeInfo and Recipe is very soon going to be very confusing. I'd rather make them be the same thing, i.e. the API is just a wrapper around the internal Recipe object. I left some ideas to that effect, maybe we can discuss it later.

File: esmvalcore/config-user.yml
Comment (stefsmeets) [reply esmvalcore/config-user.yml:40]: These are set in main by the CLI, but default values must be defined somewhere, and having them in the default user config is the most logical place.

https://github.com/ESMValGroup/ESMValCore/blob/9b6f0955d56e18f4e989435f25fe05b12d773f88/esmvalcore/_main.py#L302

File: doc/api/esmvalcore.api.recipe_info.rst
Comment (stefsmeets) [reply doc/api/esmvalcore.api.recipe_info.rst:62]: Yeah, at the moment it returns nothing. Either way works for me, but it is also a reminder that the functionality is not complete yet.

File: esmvalcore/experimental/_logging.py
Comment (stefsmeets) [reply esmvalcore/experimental/_logging.py:8]: `log_to_dir` was the best I could come up with. I updated the docstring to make it more clear what this does.

File: tests/sample_data/experimental/recipe_api_test.yml
Comment (stefsmeets) [reply tests/sample_data/experimental/recipe_api_test.yml:28]: This is already covered by the rest of the test suite.

My goal here is to have the simplest test I could come up with (suggestions welcome!) to cover the new functionality. But it does enable more complete testingusing real recipes, which is something we may want in the future.

File: setup.py
Comment (stefsmeets) [reply setup.py]: Yeah, this was temporary while debugging the CircleCI issue.

File: esmvalcore/experimental/config/_config_object.py
Comment (stefsmeets) [reply esmvalcore/experimental/config/_config_object.py:167]: The difference is that esmvalcore takes a dictionary. It calls this dictionary `config_user` (I'm not super happy with this name), which is essentially a copy of the global config (defined here: https://github.com/ESMValGroup/ESMValCore/blob/master/esmvalcore/_config.py) with the session directories added.

The same object is called `Session` in the api, but it is not a dictionary. Therefore a session must be converted to a dict for compatilibity with recipes. I considered just calling the method `.to_dict` at first, but I think `.to_config_user` makes the intent more clear.

There are two options I think:
- Use session objects everywhere (this needs master to depend on the new config, we tried this before)
- Make a compatibility layer (which is what is done in this PR)

File: esmvalcore/experimental/recipe_info.py
Comment (stefsmeets) [reply esmvalcore/experimental/recipe_info.py:78]: I specifically did not want this to depend on `config-references`, so it does not take a `TAG` directly. Tags can be resolved using the `from_tag` method.

File: esmvalcore/experimental/recipe_info.py
Comment (stefsmeets) [reply esmvalcore/experimental/recipe_info.py:189]: We will need to use `esmvalcore.Recipe` anyway to get the diagnostics / scripts, so I like the idea of inheriting from it or delegating to it. It would need a bit of refactoring to get this to work then.

 I think from a user point of view, it only makes sense to call it `Recipe` and call the `esmvalcore.Recipe` something like `RecipeDriver` or `RecipeEngine`.

File: esmvalcore/experimental/recipe_info.py
Comment (stefsmeets) [reply esmvalcore/experimental/recipe_info.py:349]: Yeah, I think this is a better solution.

File: esmvalcore/experimental/utils.py
Comment (stefsmeets) [reply esmvalcore/experimental/utils.py:97]: Good idea, I added the search locations to the `FileNotFoundError`.

**Comment by stefsmeets** [comment]
> Very nice work so far!! I had a first look at the code - didn't test anything yet. But I have a couple of suggestions and questions. My main concern is that I'm a bit worried that the distinction between RecipeInfo and Recipe is very soon going to be very confusing. I'd rather make them be the same thing, i.e. the API is just a wrapper around the internal Recipe object. I left some ideas to that effect, maybe we can discuss it later.

Hi @Peter9192 , I went through your suggestions just now. Not all of them are straightforward to answer at this point. I agree with you that `RecipeInfo` will most likely turn into `Recipe` very soon with the planned work. I have some ideas on how to go about it, but perhaps  this is something to look at together for a follow-up PR.

File: esmvalcore/experimental/recipe_info.py
Comment (Peter9192) [reply esmvalcore/experimental/recipe_info.py:78]:
```suggestion
    """Contains acknowledgement information related to this recipe.
```

File: esmvalcore/experimental/recipe_info.py
Comment (Peter9192) [reply esmvalcore/experimental/recipe_info.py:189]: Let's make a new issue to refactor the esmvalcore.Recipe class, so that it will be easier to inherit from it in this API module.

Comment (stefsmeets) [reply esmvalcore/experimental/recipe_info.py:189]: #934

**Comment by stefsmeets** [comment]
Hi @Peter9192 , have I addressed your comments? If so, could you please approve this PR?

From my point of view, this is ready to merge. But I would like to see if there is any on #936 before we do so.

**Comment by stefsmeets** [comment]
Hi @bouweandela , could you have a look at the changes to `config-user.yml` and let me know if they are in line with #936? 

And do you have any idea what is the cause of this error? It seems to be completely unrelated to the changes in the code.

```
self = <tests.unit.preprocessor._regrid.test_regrid.Test testMethod=test_regrid__cell_specification>

    def test_regrid__cell_specification(self):
        specs = ['1x1', '2x2', '3x3', '4x4', '5x5']
        scheme = 'linear'
        for spec in specs:
            result = regrid(self.src_cube, spec, scheme)
            self.assertEqual(result, self.regridded_cube)
>           self._check(spec, scheme, spec=True)

tests/unit/preprocessor/_regrid/test_regrid.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/unit/preprocessor/_regrid/test_regrid.py:23: in _check
    self.assertEqual(_CACHE[spec], self.tgt_grid)
E   AssertionError: <iris 'Cube' of unknown / (unknown) (latitude: 180; longitude: 360)> != <Mock spec='Cube' id='140365561280592'>
```

**Comment by Peter9192** [comment]
Hi @stefsmeets I thought we decided to make a PR for #934 and use it here, but maybe I'm confusing things now? I'd say the sooner we do that, the better. How do you see this now?

**Comment by stefsmeets** [comment]
> 
> 
> Hi @stefsmeets I thought we decided to make a PR for #934 and use it here, but maybe I'm confusing things now? I'd say the sooner we do that, the better. How do you see this now?

It is more necessary for the follow-up item, to get some output back. I don't see any advantages for this PR.

File: esmvalcore/experimental/recipe_info.py
Comment (Peter9192) [reply esmvalcore/experimental/recipe_info.py:189]: Didn't we agree to call this object a plain old 'Recipe' here?

**Review by Peter9192 — CHANGES_REQUESTED** [review]
I tested it now and it seems to work well. I'm fine with doing #934 after this one if you feel more comfortable with that, I just thought the earlier we do it the better, since we're still flexible.

If you could just look once more at the thing that I unresolved about the naming of the RecipeInfo class? Otherwise I'm happy with the implementation.

**Comment by Peter9192** [comment]
Oh and by the way the tests pass locally for me as well

**Comment by stefsmeets** [comment]
> I tested it now and it seems to work well. I'm fine with doing #934 after this one if you feel more comfortable with that, I just thought the earlier we do it the better, since we're still flexible.
> 
> If you could just look once more at the thing that I unresolved about the naming of the RecipeInfo class? Otherwise I'm happy with the implementation.

I think I got everything, should be good now 👍 can you press approve if you are happy? 😁

**Review by Peter9192 — APPROVED** [review]

**Comment by Peter9192** [comment]
I guess we cannot merge until codacy passes though...

**Comment by stefsmeets** [comment]
> Hi @bouweandela , could you have a look at the changes to `config-user.yml` and let me know if they are in line with #936?

After our discussion I will not define these in `config-user.yml`, I will find another location. Maybe `config_defaults.py`?

> And do you have any idea what is the cause of this error? It seems to be completely unrelated to the changes in the code.
> 
> ```
> self = <tests.unit.preprocessor._regrid.test_regrid.Test testMethod=test_regrid__cell_specification>
> 
>     def test_regrid__cell_specification(self):
>         specs = ['1x1', '2x2', '3x3', '4x4', '5x5']
>         scheme = 'linear'
>         for spec in specs:
>             result = regrid(self.src_cube, spec, scheme)
>             self.assertEqual(result, self.regridded_cube)
> >           self._check(spec, scheme, spec=True)
> 
> tests/unit/preprocessor/_regrid/test_regrid.py:107: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/unit/preprocessor/_regrid/test_regrid.py:23: in _check
>     self.assertEqual(_CACHE[spec], self.tgt_grid)
> E   AssertionError: <iris 'Cube' of unknown / (unknown) (latitude: 180; longitude: 360)> != <Mock spec='Cube' id='140365561280592'>
> ```

**Comment by stefsmeets** [comment]
This test fail seems to be completely random on CircleCI (sometime it passes, sometimes it fails), unrelated to any changed made in this PR, and I cannot reproduce it on my own machine.

> ```
> self = <tests.unit.preprocessor._regrid.test_regrid.Test testMethod=test_regrid__cell_specification>
> 
>     def test_regrid__cell_specification(self):
>         specs = ['1x1', '2x2', '3x3', '4x4', '5x5']
>         scheme = 'linear'
>         for spec in specs:
>             result = regrid(self.src_cube, spec, scheme)
>             self.assertEqual(result, self.regridded_cube)
> >           self._check(spec, scheme, spec=True)
> 
> tests/unit/preprocessor/_regrid/test_regrid.py:107: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> tests/unit/preprocessor/_regrid/test_regrid.py:23: in _check
>     self.assertEqual(_CACHE[spec], self.tgt_grid)
> E   AssertionError: <iris 'Cube' of unknown / (unknown) (latitude: 180; longitude: 360)> != <Mock spec='Cube' id='140365561280592'>
> ```

File: esmvalcore/_task.py
Comment (bouweandela) [inline esmvalcore/_task.py:40]:
```suggestion
yaml.representer.SafeRepresenter.add_representer(Path, path_representer)
yaml.representer.SafeRepresenter.add_representer(PosixPath, path_representer)
```
Why do you need to represent `CheckLevels` as yaml?

File: esmvalcore/_recipe.py
Comment (bouweandela) [inline esmvalcore/_recipe.py:309]:
```suggestion
    settings['cmor_check_data'] = dict(settings['cmor_check_metadata'])
```
I would feel safer if these were not the same dict, though it will probably work fine as is

**Review by bouweandela — CHANGES_REQUESTED** [review]
Code looks good, though I did not try using it.

Could you have a look at the question about representing CheckLevels in a YAML file? I think this should not be needed, because CheckLevels should not leak to the diagnostic script through the YAML interface, this is supposed to be used by the CMOR checker and nothing else.

Would it be possible to solve the import outside toplevel Codacy issue?

File: esmvalcore/_recipe.py
Comment (stefsmeets) [reply esmvalcore/_recipe.py:309]: Yeah good point, I was wondering about this too!

File: esmvalcore/_task.py
Comment (stefsmeets) [reply esmvalcore/_task.py:40]: Good catch, this is probably no longer needed if I do not need to have the config parameters in the `config-user.yml` 👍

**Comment by stefsmeets** [comment]
> Would it be possible to solve the import outside toplevel Codacy issue?

Yeah, it can be solved, but I think it is clearer where it is now.

**Comment by bouweandela** [comment]
> Yeah, it can be solved, but I think it is clearer where it is now.

If would be great if you could follow the style guide if there is no technical reason for deviating from it. If everyone starts making up their own style the code will become harder to read for everyone.

**Comment by stefsmeets** [comment]
> 
> 
> > Yeah, it can be solved, but I think it is clearer where it is now.
> 
> If would be great if you could follow the style guide if there is no technical reason for deviating from it. If everyone starts making up their own style the code will become harder to read for everyone.

I don't consider this error be a 'hard' fail from Codacy, sometimes the code reads much better if the import is not at the top of the file. Besides that, Codacy and yapf sometimes conflict as well, so it is unclear what style to use.

**Comment by stefsmeets** [comment]
This PR is ready to be merged. I worked around the error by clearing the cache before the regrid tests, but a better implementation is needed. For more info see #953.

**Comment by stefsmeets** [comment]
Shoot, now I merged the wrong branch here...

**Pull request merged**
